Add @implementation and feature flags for objcImpl

beccadax opened the pull request:
This PR makes a series of intertwined changes to `@_objcImplementation`:

* Adds `@implementation`, the public syntax I intend to propose
    * `@_objcImplementation` will continue to be supported for the time being; the old syntax will warn instead of error for checker diagnostics
* Adds `ObjCImplementation` experimental feature to gate access to `@implementation extension`s
    * When the feature is enabled, using `@_objcImplementation` will cause a deprecation warning
* Adds `CImplementation` experimental feature to gate access to `@implementation @_cdecl` and `@_objcImplementation @_cdecl`
    * There are very few adopters for objcImpl with cdecl yet, so I'm hoping I can get away with this retroactive change

To aid in reviewing, I'm delaying some very mechanical follow-up work until another PR:

* Renaming `ObjCImplementationAttr` to `ImplementationAttr`
* Updating comments, method names, etc. in the compiler source code
* Changing most tests from using `@_objcImplementation` to using `@implementation`

This PR will need a matching swift-syntax PR.

Fixes rdar://110728033.

beccadax committed:
Add @implementation as alias for @_objcImpl

They will eventually have slightly different deprecation status and error behavior, but not yet.

Fixes rdar://110728033.

beccadax committed:
[NFC] Add flag distinguishing @impl and @objcImpl

They will have slightly different enablement and diagnostic behavior in a future commit.

beccadax committed:
Add experimental feature flags for @implementation

• ObjCImplementation controls @implementation on extensions
• CImplementation controls @implementation and @_objcImplementation on cdecl functions

Why the difference between them? Because `@_objcImplementation extension` has already been adopted pretty widely, while `@_objcImplementation @_cdecl` is very new.

beccadax committed:
Warn that @objcImpl extension is deprecated

…only when the ObjCImplementation experimental language feature is enabled. (`@implementation extension` is the preferred form.)

beccadax committed:
Soften @impl warnings only for early adopters

Adopting @implementation turns all of the warnings into errors.

beccadax commented:
With apple/swift-syntax#2570

@swift-ci please test

beccadax committed:
Guard uses of ObjC in objc_implementation.h

Helps allow test/decl/ext/cdecl_implementation.swift to run on non-ObjC platforms.

beccadax commented:
With https://github.com/apple/swift-syntax/pull/2570

@swift-ci please test

tshortli commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1739,6 +1739,9 @@ ERROR(swift_native_objc_runtime_base_not_on_root_class,none,
       "to root classes", ())
 
+WARNING(objc_implementation_early_spelling_deprecated,none,
+        "warning-only '@_objcImplementation' spelling is deprecated; use "
```
This diagnostic is difficult for me to parse, is `warning-only` an intentional inclusion here? I would expect this to just be `'@_objcImplementation' is deprecated; use '@implementation' instead` (I don't think "spelling" is needed given that the diagnostic says "use '@implementation' instead").

beccadax replied:
There's a small functional difference between `@_objcImplementation` and `@implementation`—the former produces warnings, the latter errors—so some early adopters may need to continue using `@_objcImplementation` as a workaround for bugs for a little while longer. I was trying to nod to that in the diagnostic in case they need a reminder of why they can't update yet, but this wording probably doesn't really succeed.

beccadax replied:
Thinking about it more, I think the better thing to do is probably to only emit this warning if there aren't any other diagnostics. That way, early adopters will not get this warning until they no longer have any reason to use `@_objcImplementation`.

tshortli reviewed (approved):
LGTM! I left one suggestion for diagnostic wording.

beccadax committed:
Tweak @_objcImpl deprecation warning behavior

• Simpler wording
• Now emitted only if there are no other @objcImpl diagnostics (so @implementation would not emit any new errors)

beccadax commented:
With apple/swift-syntax#2570

@swift-ci please test

tshortli commented on test/decl/ext/cdecl_implementation_features.swift:
```diff
@@ -0,0 +1,13 @@
+
+// YES-DAG: cdecl_implementation_features.swift:[[@LINE+4]]:{{[0-9]+}}: warning: global function 'CImplFunc1' of type '(Double) -> ()' does not match type '(Int32) -> Void' declared by the header; this will become an error after adopting '@implementation'
+// NO-DAG: cdecl_implementation_features.swift:[[@LINE+3]]:{{[0-9]+}}: error: '_objcImplementation' attribute is only valid when experimental feature CImplementation is enabled
+// YES-NOT: cdecl_implementation_features.swift:[[@LINE+2]]:{{[0-9]+}}: warning: warning-only '@_objcImplementation' spelling is deprecated; use '@implementation' instead
```
This `CHECK` line looks like it might be out of date

beccadax replied:
Just noticed that myself. Good catch!

beccadax commented:
With https://github.com/apple/swift-syntax/pull/2570

@swift-ci please test

beccadax commented:
I will shortly be pushing changes that make the matching swift-syntax PR unnecessary.

beccadax committed:
Simplify and test ASTGen support

Rather than adding custom parsing to SwiftSyntax, we can parse this as a custom attribute and convert it to a built-in one in ASTGen. Test that this works correctly (and fix a bug where it wasn’t).

beccadax commented:
@swift-ci please test

nkcsgexi commented:
🎉

beccadax commented:
@swift-ci please test
